velox/exec/Task.cpp

```diff
@@ -1442,6 +1464,220 @@ Task::getLocalExchangeQueues(
+      VLOG(1) << fmt::format(
+          "{} gets {} from unused", consumerTask->taskId(), bytesForRequester);
+      return true;
+    }
+  }
+
+  std::vector<ConsumerScore> candidates;
+  available = 0;
+  for (auto [raw, weak] : consumers_) {
+    auto ptr = weak.lock();
+    if (ptr) {
+      auto otherTask = dynamic_cast<Task*>(ptr.get());
+      if (otherTask) {
+        auto size = otherTask->tracker().getCurrentTotalBytes();
+        if (size > kInitialSize) {
+          auto recoverable = otherTask->recoverableMemory();
+          if (recoverable >= kMinRecoverableBytes) {
+            candidates.push_back({ptr, otherTask, recoverable});
+            available += candidates.back().available;
+          }
+        }
+      }
+    }
+  }
+  if (available < bytesForRequester) {
+    return false;
+  }
+  std::sort(
+      candidates.begin(),
+      candidates.end(),
+      [&](const ConsumerScore& left, const ConsumerScore& right) {
+        // Most available first.
+        return left.available > right.available;
+      });
+  int64_t recovered = 0;
+  std::vector<std::shared_ptr<Task>> pausedTasks;
+  for (auto& candidate : candidates) {
+    if (auto& task = candidate.task) {
+      auto future = task->requestPause(true);
+      auto& exec = folly::QueuedImmediateExecutor::instance();
+      std::move(future).via(&exec).wait();
+      if (task->state() != TaskState::kRunning) {
+        continue;
+      }
+      VELOX_CHECK_EQ(0, task->numThreads());
+      pausedTasks.push_back(std::static_pointer_cast<Task>(candidate.consumer));
+
+      auto& taskTracker = task->tracker();
+      auto previousBytes = taskTracker.maxTotalBytes();
+      task->recover(bytesForRequester - recovered);
+      auto recoveredFromTask =
+          previousBytes - taskTracker.getCurrentTotalBytes();
+      if (recoveredFromTask < kMinRecoverableBytes) {
+        // Too little recovered. No change to limit.
+        continue;
+      }
+      recovered = roundDown(recoveredFromTask, kMinRecoverableBytes);
+      VLOG(1) << fmt::format(
+          "{} drops limit by {} to {}",
+          task->taskId(),
+          recovered,
+          taskTracker.maxTotalBytes() - recovered);
+      setLimit(taskTracker.maxTotalBytes() - recoveredFromTask, taskTracker);
+
+      recovered += recoveredFromTask;
+      if (recovered >= bytesForRequester) {
+        break;
+      }
+    }
+  }
+  bool success = recovered >= bytesForRequester;
+  // Updating downward, no throw.
+  if (success) {
+    topTracker->update(bytesForRequester - recovered);
+    VLOG(1) << fmt::format(
+        "{} increases limit by {} to {}",
+        consumerTask->taskId(),
+        bytesForRequester,
+        tracker.maxTotalBytes() + bytesForRequester);
+    setLimit(tracker.maxTotalBytes() + bytesForRequester, tracker);
+  } else {
+    topTracker->update(-recovered);
+  }
+
+  for (auto& task : pausedTasks) {
+    if (task->state() == TaskState::kRunning) {
```

Comment:
Is it thread safe? A concurrent thread can terminate the task between the state check and resume? Shall we instead to check for termination in resume method? Not sure if my understanding is correct or not. But we could leave it for followup.